### PR TITLE
Dualencoder

### PR DESCRIPTION
**Pull request by tzufgoogle**
This is a dual-encoder(two-tower) architecture that encodes both instructions and S2Cells in 
the same dense vector space. This allows us to embed new S2Cells and instructions.
The dual-encoder has a network structure for representing both S2Cells and instructions and then a and a cosine function to compute similarity between representations. 
The text encoder includes a DistilBert with uncased pretrained weights for embedding of the sentence. Then the CLS token embedding goes through a feed-forward layer, a RELU activation function and another feed-forward layer to reduce the dim to X.
The cellid is a binary vector that goes through one feed-forward layer, such that the dimension changes to X, the same as the text representation dimension. We train with three cellids: The correct cellid, a neighbor cellid, and a far cellid (more than 2 KM). We sum the losses over the three cellids and backpropagate.

### CODE REVIEW

File: cabby/model/text/util.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  binary_range = 2**np.arange(dim, dtype=np.uint64)
+  vector = array_int.reshape(-1,1)
+
+  bin_rep = ((vector & binary_range) != 0).astype(int)
+  return bin_rep
+
+
+def save_checkpoint(save_path: Text, model:  torch.nn.Module,
+          valid_loss: float):
+  '''Funcion for saving model.'''
```
[the remark this answers is not in the document]
Comment (jasonbaldridge) [reply cabby/model/text/util.py:37]: also, several times below

File: cabby/model/text/util.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    return
+
+  state_dict = torch.load(load_path, map_location=device)
+  logging.info(f'Results loaded from <== {load_path}')
+
+  return state_dict
+
+
+def predictions_to_points(preds: Sequence,
+              label_to_cellid: Dict[int, int]) -> Sequence[Tuple[float, float]]:
```
Comment (jasonbaldridge) [inline cabby/model/text/util.py:98]: Fix indentation.

**Review by jasonbaldridge — CHANGES_REQUESTED** [review]
Great to have the dual encoder implementation! I have a bunch of bigger and smaller requests. Please go ahead and resolve the conversations for the trivial stuff when you've addressed them.

I'm still confused as to why your editor isn't sorting out the formatting stuff correctly. I'd appreciate it if you could look into it and make it work. That would save both of us time and mental effort in future.

File: cabby/model/text/util.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+from absl import logging
+import numpy as np
+from sklearn.metrics import accuracy_score
+import torch
+
+from cabby.geo import util
+
+
+def binary_representation(array_int, dim):
```
[the remark this answers is not in the document]
Comment (tzufgoogle) [reply cabby/model/text/util.py]: Added both doc and test

File: cabby/model/text/util.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  binary_range = 2**np.arange(dim, dtype=np.uint64)
+  vector = array_int.reshape(-1,1)
+
+  bin_rep = ((vector & binary_range) != 0).astype(int)
+  return bin_rep
+
+
+def save_checkpoint(save_path: Text, model:  torch.nn.Module,
+          valid_loss: float):
```
[the remark this answers is not in the document]
Comment (tzufgoogle) [reply cabby/model/text/util.py]: fixed

File: cabby/model/text/util.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tzufgoogle) [reply cabby/model/text/util.py]: Fixed all

File: cabby/model/text/util.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+  state_dict = {'model_state_dict': model.state_dict(),
+          'valid_loss': valid_loss}
+
+  torch.save(state_dict, save_path)
+  logging.info(f'Model saved to ==> {save_path}')
+
+
+def load_checkpoint(load_path: Text, model:  torch.nn.Module,
+          device: torch.device) -> Dict[Text, Sequence]:
```
[the remark this answers is not in the document]
Comment (tzufgoogle) [reply cabby/model/text/util.py]: fixed.

File: cabby/model/text/util.py (hunk 2 above)
Comment (tzufgoogle) [reply cabby/model/text/util.py]: Fixed.

**Review by tzufgoogle — COMMENTED** [review]
Made the requested changes including adding TextGeoDataset

File: cabby/model/text/dual_encoder/dataset.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+    self.cellids = util.binary_representation(cellids_array, dim = CELLID_DIM)
+    self.neighbor_cells =  util.binary_representation(
+      neighbor_cells_array, dim = CELLID_DIM)
+
+    self.far_cells =  util.binary_representation(
+      far_cells_array, dim = CELLID_DIM)
+
+    
+  def __getitem__(self, idx: int):
```
[the remark this answers is not in the document]
Comment (jasonbaldridge) [reply cabby/model/text/dual_encoder/dataset.py:82]: Thanks!

File: cabby/model/text/dual_encoder/model_trainer.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+  dataset_path = os.path.join(FLAGS.dataset_dir, str(FLAGS.s2_level))
+  train_path_dataset = os.path.join(dataset_path,'train.pth')
+  valid_path_dataset = os.path.join(dataset_path,'valid.pth')
+  unique_cellid_path = os.path.join(dataset_path,"unique_cellid.npy")
+  tesnor_cellid_path = os.path.join(dataset_path,"tensor_cellid.pth")
+  label_to_cellid_path = os.path.join(dataset_path,"label_to_cellid.npy")
+
+
+  if os.path.exists(dataset_path):
```
[the remark this answers is not in the document]
Comment (jasonbaldridge) [reply cabby/model/text/dual_encoder/model_trainer.py:96]: Great! Minor nit: seems like we don't need named args for loading and saving, but no problem if you want to keep it as is.

File: cabby/model/text/dual_encoder/train.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  # Validation loop.
+  logging.info("Starting evaluation.")
+  true_points_list, pred_points_list, predictions_list, true_vals = [], [], [], []
+  correct = 0
+  total = 0
+  loss_val_total = 0
+
+  for batch in valid_loader:
+    text, cellids, neighbor_cells, far_cells, points, labels = batch
+    cellids, neighbor_cells, far_cells = cellids.float().to(device), neighbor_cells.float().to(device), far_cells.float().to(device)
```
[the remark this answers is not in the document]
Comment (jasonbaldridge) [reply cabby/model/text/dual_encoder/train.py:59]: Doesn't seem changed. Did you forget to push the commit?

**Review by jasonbaldridge — APPROVED** [review]
Thanks! LGTM. ( But you might still need to push a missing commit?)

**Comment by sayalikulkarni** [comment]
Great to see this out! Thanks Jason for covering up the review.
